Add sqlite3 database support

hermankolden opened the pull request:
Since ULog is a streamed format, large files (>20 mB) take a long time to load from disk. This is a problem every time such large logs are loaded, and furthermore prohibits the comparison of many logs for e.g. statistical purposes.

This commit solves these two problems by introducing a class, inheriting from ULog, that serializes the ulog format into database tables stored in a sqlite3 database.

The following code example shows the difference in load times:

```
from pyulog import ULog
from pyulog.db import DatabaseULog
from timeit import timeit

db_handle = DatabaseULog.get_db_handle('ulog.sqlite3')

res={'ulog': None, 'primary_key': None}
def load_ulog():
    res['ulog'] = ULog('test/sample_log_medium.ulg')

def save_to_db():
    dbulog = DatabaseULog(db_handle, ulog=res['ulog'])
    dbulog.save()
    res['primary_key'] = dbulog.primary_key

def load_dbulog():
    dbulog = DatabaseULog(db_handle, primary_key=res['primary_key'])

print(timeit(load_ulog, number=1)) # 1.1072970120003447
print(timeit(save_to_db, number=1)) # 0.08976522006560117
print(timeit(load_dbulog, number=1)) # 0.014600005000829697
```

bkueng commented on pyulog/db.py:
```diff
@@ -0,0 +1,572 @@
+    Another weakness of the class is that there are some silently failing
+    states if you don't call save() immediately after instantiating from a ULog
+    object. The requirement of explicit save() is to prevent unexpected,
+    suddent side effects, which is considered worse than the current solution.
```
```suggestion
    sudden side effects, which is considered worse than the current solution.
```

bkueng commented on pyulog/db.py:
```diff
@@ -0,0 +1,572 @@
+        Note that the parameters name and multi_instance parameters are named
+        the same way as the original ULog file for compatibility.
+        '''
+        with self._db() as con:
```
Is opening the db connection cheap?

hermankolden replied:
In general no, but I'm not sure if I understand the intent behind the question.

bkueng replied:
Depending on the use-case, this method might be called a lot, so I'd make sure to keep the connection open.

hermankolden replied:
Storing an open connection would require always opening/closing when using DatabaseULog, which is pretty cumbersome. Instead I added an option of supplying a cursor directly if one is doing a lot of get_dataset. Updated PR coming soon.

bkueng commented on test/test_db.py:
```diff
@@ -0,0 +1,246 @@
+
+    The following attributes are not tested with the current suite, since they
+    are not contained in the tested ulogs:
+        - has_data_appended
+        - file_corruption
+        - _default_parameters
+        - _changed_parameters
+        - logged_messages_tagged
```
For most of these there are sample logs

hermankolden replied:
Great! I'll add those

bkueng commented on test/test_db.py:
```diff
@@ -0,0 +1,246 @@
+                    self.assertTrue(np.array_equal(array1, array2, equal_nan=True))
+            )
+
+    def test_medium_log(self):
```
You can use
```suggestion
    @data('sample', 'sample_appended', 'sample_appended_multiple', 'sample_logging_tagged_and_default_params')
    def test_log(self, base_name):
```

bkueng commented on pyulog/db.py:
```diff
@@ -0,0 +1,572 @@
+        self._pk = primary_key
+        self._db = db_handle
+
+        attrs = [
+            ('_debug', False),
+            ('_file_corrupt', False),
+            ('_start_timestamp', 0),
+            ('_last_timestamp', 0),
+            ('_msg_info_dict', {}),
+            ('_msg_info_multiple_dict', {}),
+            ('_initial_parameters', {}),
+            ('_default_parameters', {}),
+            ('_changed_parameters', []),
+            ('_message_formats', {}),
+            ('_logged_messages', []),
+            ('_logged_messages_tagged', {}),
+            ('_dropouts', []),
+            ('_data_list', []),
+            ('_subscriptions', {}),
+            ('_filtered_message_ids', set()),
+            ('_missing_message_ids', set()),
+            ('_file_version', 0),
+            ('_compat_flags', [0] * 8),
+            ('_incompat_flags', [0] * 8),
+            ('_appended_offsets', []),
+            ('_has_sync', True),
+            ('_sync_seq_cnt', 0),
+        ]
+        ULog._disable_str_exceptions = disable_str_exceptions
+
+        if ulog is not None:
+            for attr_name, _ in attrs:
+                setattr(self, attr_name, getattr(ulog, attr_name))
+        else:
+            for attr_name, default_value in attrs:
+                setattr(self, attr_name, default_value)
+            self.load()
```
To avoid this I'd rather make the `log_file` optional in the `ULog` initializer.

bkueng reviewed (commented):
Looks very clean!

Can you look at the few remaining CI issues?

Is the `test/sample_log_medium.ulg` file primarily for speed comparison? I'm a bit hesitant to add larger binary files to the repo.

> Since ULog is a streamed format, large files (>20 mB) take a long time to load from disk.

What I found is that it's really just python that's very inefficient at reading & parsing binary files.

hermankolden commented:
@bkueng I've fixed your comments now, along with some issues that I discovered in the process.

Database setup required running ".read sql/setup_db.sql" manually, which was unintuitive when using pyulog as a library. I hence added the ulog_migratedb console script to handle the database setup.

When handling the database setup issue, I realized a single SQL script isn't robust to future database changes. If there is a discovered issue with the current database schema (like the missing fields for changed_parameters etc that were discovered when I added more tests), or there is a bug in the schema, then new changes might be needed, and we really don't want to lose data by accident. Hence, careful data migration might be needed.

I researched some migration tools like "alembic" and "migrations", but the former was sqlalchemy-specific and the latter seemed to have little support. I hence created a small migration script, migrate_db.py, which handles simple forward-migration. I do worry that this might be overengineered, but the migration logic itself is only 50 lines or so, so IMO it's worth the future-proofing.

hermankolden commented:
Oops, I didn't mean to close

bkueng reviewed (commented):
Generally looks really good, just a couple of details.
Do you still want to add `test/sample_log_medium.ulg`?
Can you also cleanup the commits?
